### ISSUE TITLE
Different syntax

### ISSUE THREAD

**Issue by ghost** [bamless/jstar#17]
I know it's probably to late for something like this, but I don't like the lua (I think) like syntax, with then do, then, and end keywords. I feel like it adds to many keywords, and that it would be better if it used braces, or a pythonic syntax. If I could I'd help change it but I'm not sure if you agree with me and I'm also am very bad with c.

edit: I did bad job explaining what I want. I don't like how lua uses different keywords for loops and if statements, that "do", and "then" keywords. I'd prefer if it used a single keyword

**Comment by bamless** [comment]
Honestly, I've been thinking about making the  `then` and `do` keywords optional for a while.
Let me explain better; since the language already uses newlines as statement ending characters, we can do something like this:
```lua
if cond
    ...
end
```
in this instance the parser would see this:
```lua
TOK(if) TOK(cond) TOK('\n') <-- This will take the place of 'then' while parsing
    ...
TOK(end)
```
And use the '\n' in place of `then` for the start of the block.
Same goes for `while`/ `for` and the `do` keyword.

Now, this solution would work well, but I still want to mantain support for the old syntax, both for backwards-compatibility and because it permits writing one-line statements:
```lua
if cond then print('true!') end
```
Without `then` the parser would have no idea when to stop parsing the condition and start the block.
Unfortunately this means that the old keywords would still be reserved.

As for the `end` keyword, well, this is unfortunately necessary.  
The only way to get rid of it would be to use braces (which I don't want to do at this stage of the language), or by using indentation-style blocks like in python, which would add other classes of problems in the language at large.

Having said that, if you're intrested in this change i would gladly implement it.
I still need to figure a few things like how to restructure the CLI to permit multi-line input without relying on the `then` or `do` keyword to signal the start of a block, but aside from this should be pretty simple.

**Comment by ghost** [comment]
I don't have much of a problem with the end keyword at all, I really like this solution

**Comment by bamless** [comment]
I sort of implemented the change in the opt-keywords branch. You can take a look at the [core.jsr](https://github.com/bamless/jstar/blob/opt-keywords/src/std/core.jsr) file to see how the language looks without the `then` and `do` keyword.

Honestly, the change looked good on paper but without the keywords the code is a bit harder to read in my opinion, I'm not quite sure if I want to merge it into the main branch at this point. I need to think this through.

**Comment by ghost** [comment]
I looked at it, It does seem hard to read. I also didn't think about how it made the language more dependent on white space. I think the way you made the language was better.

**Comment by bamless** [comment]
I decided to take a look at how other languages handle the `then` and `do` keywords. As I was searching I stumbled upon Julia, that simply omits them and doesn't seem to rely on newlines to tell the start of a block. It made me realize that, since in J* we do not allow side-effect-free expressions to appear as statements, we don't need to check for newlines too! Code like this is actually possible:
```lua
if condition print("true!") end
```
As a bonus, omitting the keywords in this way, actually simplifies the implementation of the parser a fair bit!
Also, after some time getting used to the syntax I changed my opinion on it, and think this way the code is more pleasant to write and more approachable to newcomers. 
So, in the end I completely stripped the keywords from the grammar, making it more simple and allowing the use of `then` and `do` as variable names. 
You can check out the changes in version [1.7](https://github.com/bamless/jstar/releases/tag/v1.7)

**Comment by bamless** [comment]
Implemented in f49d72ca6af361fc8550031d45be5b196821be92
